runtime/src/main/scala/com/armanbilge/vilcacora/runtime/Interpreter.scala

```diff
@@ -0,0 +1,305 @@
+
+import cats.effect.{IO, Resource}
+import cats.syntax.all._
+import com.armanbilge.vilcacora.ir._
+import com.armanbilge.vilcacora.runtime.LibSvm._
+import scala.scalanative.unsafe._
+import scala.scalanative.libc.stdlib
+import scala.scalanative.libc.string.memcpy
+import scala.scalanative.unsigned._
+
+/** The core execution engine for a translated `ModelIR`. It manages memory and executes operations
+  * within a Cats Effect IO context.
+  */
+object Interpreter {
+
+  // A type alias for clarity: maps tensor names to their C memory pointers.
+  type MemoryMap = Map[String, Ptr[Byte]]
+
+  /** Executes a complete ModelIR graph from inputs to outputs. */
+  def execute(model: ModelIR, inputs: Map[String, Array[_]]): IO[Map[String, Array[_]]] =
+    // `Resource.use` guarantees all tensor memory is pre-allocated and safely freed.
+    memoryResource(model).use { memoryMap =>
+      for {
+        _ <- copyInputsToMemory(inputs, model, memoryMap)
+
+        // The core execution loop. It executes each operation and then yields control
+        // to the Cats Effect runtime to maintain application responsiveness.
+        // This is the desired structure.
+        _ <- model.operations.traverse_ { op =>
+          executeOperation(op, memoryMap, model) <* IO.cede
+        }
+
+        results <- copyOutputsFromMemory(model, memoryMap)
+      } yield results
+    }
+
+  /** A `Resource` that manages the lifecycle of all tensor memory for the graph. */
+  private def memoryResource(model: ModelIR): Resource[IO, MemoryMap] = {
+    val allocAll = IO {
+      model.allocations.map { case (name, allocation) =>
+        val totalBytes = (allocation.shape.product * allocation.dataType.sizeInBytes).toUSize
+        val ptr = stdlib.malloc(totalBytes)
+        if (ptr == null) throw new OutOfMemoryError(s"Failed to allocate tensor '$name'")
+        allocation.initialData.foreach(data =>
+          memcpy(ptr, data.at(0).asInstanceOf[Ptr[Byte]], data.length.toUSize),
+        )
+        name -> ptr
+      }
+    }
+    Resource.make(allocAll)(memoryMap => IO(memoryMap.values.foreach(stdlib.free))).map(_.toMap)
+  }
+
+  /** Executes a single operation by dispatching to the appropriate handler. */
+  private def executeOperation(op: Operation, memory: MemoryMap, model: ModelIR): IO[Unit] =
+    op match {
+      case op: Operation.SVMClassifier => handleSvmClassifier(op, memory, model)
+      case op: Operation.Add => handleElementWise(op, memory, model)(_ + _)
+      case op: Operation.Mul => handleElementWise(op, memory, model)(_ * _)
+      case op: Operation.Cast => handleCast(op, memory, model)
+      case other =>
+        IO.raiseError(
+          new NotImplementedError(s"Operation not implemented: ${other.getClass.getSimpleName}"),
+        )
+    }
+
+  // --- Operation Handlers ---
+
+  private def handleElementWise(op: Operation, memory: MemoryMap, model: ModelIR)(
+      f: (Float, Float) => Float,
+  ): IO[Unit] = IO {
```

Comment:
Unfortunate the `f: (Float, Float) => Float` will be allocated as a lambda and use dynamic dispatch. The Scala Native Interflow optimizer may be able to specialize this and avoid the lambda allocation and dispatch, but unfortunately it would be better to write these as separate methods with the operation hard-coded, for the guaranteed best performance.